File: app/src/main/java/com/melvinhou/dimension2/media/tiktok/TiktokActivity.java
package com.melvinhou.dimension2.media.tiktok;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.TextUtils;
import android.text.style.BackgroundColorSpan;
import android.text.style.ForegroundColorSpan;
import android.text.style.RelativeSizeSpan;
import android.text.style.URLSpan;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.RequestOptions;
import com.melvinhou.dimension2.Comment;
import com.melvinhou.dimension2.Dimension2Application;
import com.melvinhou.dimension2.R;
import com.melvinhou.dimension2.databinding.ActTikTokBD;
import com.melvinhou.kami.mvp.MvpActivity2;
import com.melvinhou.kami.util.FcUtils;
import com.melvinhou.kami.util.ResourcesUtils;

import java.math.BigDecimal;

import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.PagerSnapHelper;

/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2021/4/21 20:19
 * <p>
 * = 分 类 说 明：模仿抖音
 * ================================================
 */
public class TiktokActivity extends MvpActivity2<TiktokCotract.Presenter, ActTikTokBD> implements TiktokCotract.View {

    private TikTokAdapter mAdapter;

    @Override
    protected int getLayoutID() {
        return R.layout.activity_tiktok;
    }

    @Override
    protected void initWindowUI() {
//        getWindow().setBackgroundDrawableResource(R.color.black);
        getWindow().getDecorView().setBackgroundColor(Color.BLACK);
        getWindow().getDecorView().setSystemUiVisibility(
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
        getWindow().setStatusBarColor(Color.TRANSPARENT);
        getWindow().setNavigationBarColor(Color.BLACK);
    }

    @Override
    protected TiktokCotract.Presenter upPresenter() {
        ViewModelProvider.Factory factory = ViewModelProvider.AndroidViewModelFactory
                .getInstance(Dimension2Application.getInstance());
        ViewModelProvider viewModelProvider = new ViewModelProvider(this, factory);
        TiktokModel model = viewModelProvider.get(TiktokModel.class);
        return new TikTokPresenter(this, model);
    }

    @Override
    protected void initView() {

    }

    @Override
    protected void initListener() {
        getViewDataBinding().list.setLayoutManager(new LinearLayoutManager(
                FcUtils.getContext(), LinearLayoutManager.VERTICAL, false) {
            @Override
            public boolean canScrollVertically() {
                return getPresenter().canScrollVertically(super.canScrollVertically());
            }
        });
        new PagerSnapHelper().attachToRecyclerView(getViewDataBinding().list);
        mAdapter = new TikTokAdapter();
        getViewDataBinding().list.setAdapter(mAdapter);
        getViewDataBinding().list.addOnScrollListener(getPresenter().getListScrollListener());
        getViewDataBinding().comments.setOnClickListener(getPresenter()::openCommentDrawer);
        getViewDataBinding().commentRoot.setStateListener(getPresenter()::onChangeUnfoldState);
        getViewDataBinding().show.setOnClickListener(getPresenter()::closeCommentDrawer);
        mAdapter.setOnItemDoubleClickListener(getPresenter().getItemDoubleClickListener());
    }


    @SuppressLint("CheckResult")
    @Override
    protected void initData() {

    }


    @Override
    public TikTokAdapter getListAdapter() {
        return mAdapter;
    }

    @Override
    public void setCurrentData(TiktokEntity entity) {
        getViewDataBinding().setEntity(entity);
        //手动优化
        getViewDataBinding().star.setText(formatNumber(entity.getStarCount()));
        getViewDataBinding().comments.setText(formatNumber(entity.getCommentCount()));
        getViewDataBinding().share.setText(formatNumber(entity.getShareCount()));
        getViewDataBinding().commentCount.setText(
                new StringBuffer("共")
                        .append(formatNumber(entity.getCommentCount()))
                        .append("条评论"));
    }

    /**
     * 矫正数据显示
     *
     * @param number
     */
    private String formatNumber(int number) {
        String str = String.valueOf(number);
        if (number > 9999) {
            float num1 = ((float) number) / 10000f;
            //保留两位小数
            str = new BigDecimal(num1)
                    .setScale(2, BigDecimal.ROUND_HALF_UP)
                    .floatValue()
                    + "w";
        }
        return str;
    }

    @Override
    public void switchItemCanTouchState(boolean value) {
        getViewDataBinding().show.setCanTouch(!value);
    }

    @Override
    public void closeCommentDrawer() {
        getViewDataBinding().commentRoot.close();
    }

    @Override
    public void openCommentDrawer() {
        getViewDataBinding().commentRoot.open();
    }

    @Override
    public void updateStarCount(int starCount) {
        getViewDataBinding().star.setText(formatNumber(starCount));
    }

    @Override
    public void showStarAnimation(float x, float y) {
        getViewDataBinding().show.show(x, y);
    }

    @Override
    public void clearCommentItems() {
        getViewDataBinding().commentsContainer.removeAllViews();
    }

    @Override
    public void jionHigCommentItem(Comment comment) {
        View view = View.inflate(FcUtils.getContext(), R.layout.item_comment_hig, null);
        getViewDataBinding().commentsContainer.addView(view);
        ImageView photoView = view.findViewById(R.id.user_photo);
        TextView titleView = view.findViewById(R.id.user_name);
        TextView contextView = view.findViewById(R.id.context);
        Glide.with(photoView)
                .load(comment.getUser().getPhoto())
                .apply(new RequestOptions()
                        .override(240, 240)
                        .placeholder(R.mipmap.fc)
                        .error(R.mipmap.fc)
                        .diskCacheStrategy(DiskCacheStrategy.ALL))
                .into(photoView);
        titleView.setText(comment.getUser().getNickName());
        contextView.setText(comment.getContent());
    }

    @Override
    public void jionSubCommentItem(Comment.SubComment comment) {
        View view = View.inflate(FcUtils.getContext(), R.layout.item_comment_sub, null);
        getViewDataBinding().commentsContainer.addView(view);
        ImageView photoView = view.findViewById(R.id.user_photo);
        TextView titleView = view.findViewById(R.id.user_name);
        TextView contextView = view.findViewById(R.id.context);
        Glide.with(photoView)
                .load(comment.getUser().getPhoto())
                .apply(new RequestOptions()
                        .override(120, 120)
                        .placeholder(R.mipmap.fc)
                        .error(R.mipmap.fc)
                        .diskCacheStrategy(DiskCacheStrategy.ALL))
                .into(photoView);
        //用户名

        SpannableStringBuilder spbuilder = new SpannableStringBuilder();
        if (comment.getUser().getNickName().equals(getViewDataBinding().getEntity().getUser().getNickName())){
            spbuilder.append("\r作者\r");
            spbuilder.setSpan(new ForegroundColorSpan(Color.WHITE),
                    0, spbuilder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            spbuilder.setSpan(new BackgroundColorSpan(Color.RED),
                    0, spbuilder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            spbuilder.setSpan(new RelativeSizeSpan(0.8f),
                    0, spbuilder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            spbuilder.append("\r\r");
        }
        titleView.setText(spbuilder.append(comment.getUser().getNickName()));

        //评论内容
        SpannableStringBuilder builder = new SpannableStringBuilder();
        if (!TextUtils.isEmpty(comment.getCorrelativeUserName())) {
            int size = 0;
            builder.append("回复").append("\r");
            builder.setSpan(new ForegroundColorSpan(ResourcesUtils.getColor(R.color.gray)),
                    size, builder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            size = builder.length();
            builder.append("@").append(comment.getCorrelativeUserName());
            builder.setSpan(new URLSpan("https://www.baidu.com/"),
                    size, builder.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.append(":\r\r");
        }
        contextView.setText(builder.append(comment.getContent()));
    }

    @Override
    public void back() {
        getPresenter().back();
    }
}

File: tiktok-sample/src/main/java/com/melvinhou/tiktok_sample/TiktokActivity.kt
package com.melvinhou.tiktok_sample

import android.graphics.Color
import android.net.Uri
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.TextUtils
import android.text.style.BackgroundColorSpan
import android.text.style.ForegroundColorSpan
import android.text.style.RelativeSizeSpan
import android.text.style.URLSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.view.isVisible
import androidx.viewpager2.widget.ViewPager2
import com.melvinhou.kami.adapter.BindRecyclerAdapter
import com.melvinhou.kami.mvp.MvpActivity2
import com.melvinhou.knight.loadImage
import com.melvinhou.medialibrary.video.FcVideoView
import com.melvinhou.tiktok_sample.bean.Comment
import com.melvinhou.tiktok_sample.bean.TiktokEntity
import com.melvinhou.tiktok_sample.databinding.ActivityTiktok2Binding
import com.melvinhou.tiktok_sample.databinding.ItemTiktok2Binding


/**
 * ===============================================
 * = 作 者：风 尘
 * <p>
 * = 版 权 所 有：melvinhou@163.com
 * <p>
 * = 地 点：中 国 北 京 市 朝 阳 区
 * <p>
 * = 时 间：2023/5/16 0016 17:29
 * <p>
 * = 分 类 说 明：mvp-v
 * ================================================
 */
class TiktokActivity : MvpActivity2<ActivityTiktok2Binding, TiktokCotract.Presenter>(),
    TiktokCotract.View {

    override fun upPresenter(): TiktokCotract.Presenter = TiktokPresenter(this)

    override fun openViewBinding(): ActivityTiktok2Binding =
        ActivityTiktok2Binding.inflate(layoutInflater)


    private lateinit var adapter: MyAdapter

    //vp的回调
    private val mPageChangeCallback = object : ViewPager2.OnPageChangeCallback() {
        override fun onPageSelected(position: Int) {
            adapter.getData(position)?.let {
                binding.inInfo.tvTitle.text = "@${it.user.nickName}"
                binding.inInfo.tvExplain.text = it.title
                //评论
                clearCommentItems()
                it.comments.forEach { comment ->
                    jionHigCommentItem(comment)
                    comment.subComments.forEach { subComment ->
                        jionSubCommentItem(subComment)
                    }
                }
            }
        }
    }

    override fun initView() {
        binding.container.isUserInputEnabled = true//用户输入
        binding.container.offscreenPageLimit = ViewPager2.OFFSCREEN_PAGE_LIMIT_DEFAULT//预加载
        //适配器
        adapter = MyAdapter()
        binding.container.adapter = adapter
    }

    override fun initListener() {
        adapter.setOnItemClickListener { viewHolder, position, data ->
            if (binding.inComment.root.isOpen)
                binding.inComment.root.close()
            else
                viewHolder.binding.video.apply {
                    if (isPlaying) {
                        pause()
                    } else {
                        start()
                    }
                }
        }
        binding.inComment.root.setStateListener {
            binding.container.isUserInputEnabled = !it//用户输入
        }
        binding.inTools.btComment.setOnClickListener {
            binding.inComment.root.open()
        }
    }

    override fun onResume() {
        super.onResume()
        binding.container.registerOnPageChangeCallback(mPageChangeCallback)
    }

    override fun onPause() {
        super.onPause()
        binding.container.unregisterOnPageChangeCallback(mPageChangeCallback)
    }

    override fun addItems(isRefresh: Boolean, items: ArrayList<TiktokEntity>) {
        if (isRefresh) adapter.clearData()
        adapter.addDatas(items)
    }


    //清空评论
    private fun clearCommentItems() {
        binding.inComment.commentsContainer.removeAllViews()
    }

    //添加评论
    private fun jionHigCommentItem(comment: Comment) {
        val view = View.inflate(this, R.layout.item_comment_hig, null)
        binding.inComment.commentsContainer.addView(view)
        val photoView = view?.findViewById<ImageView>(R.id.user_photo)
        val titleView = view?.findViewById<TextView>(R.id.user_name)
        val contextView = view?.findViewById<TextView>(R.id.context)
        photoView?.loadImage(comment.user.photo)
        titleView?.setText(comment.user.nickName)
        contextView?.setText(comment.content)
    }

    //添加子评论
    private fun jionSubCommentItem(comment: Comment.SubComment) {
        val view = View.inflate(this, R.layout.item_comment_sub, null)
        binding.inComment.commentsContainer.addView(view)
        val photoView = view?.findViewById<ImageView>(R.id.user_photo)
        val titleView = view?.findViewById<TextView>(R.id.user_name)
        val contextView = view?.findViewById<TextView>(R.id.context)
        photoView?.loadImage(comment.user.photo)
        titleView?.setText(comment.user.nickName)
        contextView?.setText(comment.content)
        //用户名
        val currentUser = adapter.getData(binding.container.currentItem)
        val spbuilder = SpannableStringBuilder()
        if (comment.user.nickName == currentUser.user.nickName) {
            spbuilder.append("\r作者\r")
            spbuilder.setSpan(
                ForegroundColorSpan(Color.WHITE),
                0, spbuilder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spbuilder.setSpan(
                BackgroundColorSpan(Color.RED),
                0, spbuilder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spbuilder.setSpan(
                RelativeSizeSpan(0.8f),
                0, spbuilder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spbuilder.append("\r\r")
        }
        titleView?.setText(spbuilder.append(comment.user.nickName))

        //评论内容
        val builder = SpannableStringBuilder()
        if (!TextUtils.isEmpty(comment.getCorrelativeUserName())) {
            var size = 0
            builder.append("回复").append("\r")
            builder.setSpan(
                ForegroundColorSpan(getColor(R.color.gray)),
                size, builder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            size = builder.length
            builder.append("@").append(comment.getCorrelativeUserName())
            builder.setSpan(
                URLSpan("https://www.baidu.com/"),
                size, builder.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            builder.append(":\r\r")
        }
        contextView?.setText(builder.append(comment.getContent()))
    }


    internal class MyAdapter :
        BindRecyclerAdapter<TiktokEntity, ItemTiktok2Binding>() {
        override fun getViewBinding(
            inflater: LayoutInflater,
            parent: ViewGroup
        ): ItemTiktok2Binding {
            return ItemTiktok2Binding.inflate(inflater, parent, false)
        }

        override fun bindData(
            binding: ItemTiktok2Binding,
            position: Int,
            data: TiktokEntity
        ) {
            binding.videoLoading.isVisible = true
            binding.videoErrorMessage.isVisible = false
            binding.videoPlay.isVisible = false
            binding.video.setVideoURI(Uri.parse(data.url))
            //监听
            binding.video.setOnPreparedListener {
                binding.videoLoading.isVisible = false
            }
            binding.video.setOnErrorListener { mp, what, extra ->
                binding.videoErrorMessage.isVisible = true
                binding.videoErrorMessage.text = "视频错误"
                true
            }
            binding.video.setPlayerStateListener {
                when (it) {
                    FcVideoView.STATE_PAUSED,
                    FcVideoView.STATE_STOPPED,
                    FcVideoView.STATE_PLAYBACK_COMPLETED,
                    -> {
                        binding.videoPlay.isVisible = true
                    }
                    else -> {
                        binding.videoPlay.isVisible = false
                    }
                }
            }
        }
    }
}